go/libraries/doltcore/diff/querydiff/sort_node.go

```diff
@@ -0,0 +1,252 @@
+	fromIter, err := from.RowIter(fromCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	toIter, err := to.RowIter(toCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	ae := atomicerr.New()
+
+	return &sortNodeDiffer{
+		fromNode: from,
+		toNode:   to,
+		fromIter: newIterQueue(fromIter, ae),
+		toIter:   newIterQueue(toIter, ae),
+		lastCmp:  unknown,
+		ae:       ae,
+	}, nil
+}
+
+type sortNodeDiffer struct {
+	// todo: do we need two contexts?
+	ctx      *sql.Context
+	fromNode *plan.Sort
+	toNode   *plan.Sort
+	fromIter *iterQueue
+	toIter   *iterQueue
+	lastCmp  rowCmp
+	ae       *atomicerr.AtomicError
+}
+
+var _ NodeDiffer = &sortNodeDiffer{}
+
+func (nd *sortNodeDiffer) nextFromRow() (sql.Row, error) {
+	nd.fromIter.maybeStart()
+	nd.toIter.maybeStart()
+
+	if nd.fromIter.isDone() {
+		return nil, io.EOF
+	}
+	if nd.toIter.isDone() {
+		return nd.fromIter.pop(), nil
+	}
+
+	if nd.lastCmp != unknown {
+		panic("query diff iterators called out of order")
+	}
+
+	var err error
+	nd.lastCmp, err = nd.rowCompare(nd.fromIter.peek(), nd.toIter.peek())
+	if err != nil {
+		return nil, err
+	}
+
+	switch nd.lastCmp {
+	case lesser:
+		return nd.fromIter.pop(), nil
+	case equal:
+		return nd.fromIter.pop(), nil
+	case greater:
+		return nil, errSkip
+	default:
+		panic("incorrect value fromIter rowCmp")
+	}
+}
+
+func (nd *sortNodeDiffer) nextToRow() (sql.Row, error) {
+	nd.fromIter.maybeStart()
+	nd.toIter.maybeStart()
+
+	if nd.toIter.isDone() {
+		return nil, io.EOF
+	}
+	if nd.fromIter.isDone() && nd.lastCmp == unknown {
+		return nd.toIter.pop(), nil
+	}
+	// if lastCmp != unknown, fromIter just popped its last item
+
+	if nd.lastCmp == unknown {
+		panic("query diff iterators called out of order")
+	}
+
+	cmp := unknown
+	nd.lastCmp, cmp = cmp, nd.lastCmp
+
+	switch cmp {
+	case lesser:
+		return nil, errSkip
+	case equal:
+		return nd.toIter.pop(), nil
+	case greater:
+		return nd.toIter.pop(), nil
+	default:
+		panic("incorrect value fromIter rowCmp")
+	}
+}
+
+func (nd *sortNodeDiffer) rowCompare(left, right sql.Row) (rowCmp, error) {
+	if left == nil || right == nil {
+		panic("nil rows cannot be compared")
+	}
+
+	for _, sf := range nd.fromNode.SortFields {
+		typ := sf.Column.Type()
+		lv, err := sf.Column.Eval(nd.ctx, left)
+		if err != nil {
+			return unknown, err
+		}
+
+		rv, err := sf.Column.Eval(nd.ctx, right)
+		if err != nil {
+			return unknown, err
+		}
+
+		if sf.Order == plan.Descending {
+			lv, rv = rv, lv
+		}
+
+		if lv == nil && rv == nil {
+			continue
+		} else if lv == nil {
+			if sf.NullOrdering == plan.NullsFirst {
+				return lesser, nil
+			} else {
+				return greater, nil
+			}
+		} else if rv == nil {
+			if sf.NullOrdering == plan.NullsFirst {
+				return greater, nil
+			} else {
+				return lesser, nil
+			}
+		}
+
+		cmp, err := typ.Compare(lv, rv)
+		if err != nil {
+			return 0, err
+		}
+		if cmp != 0 {
+			return rowCmp(cmp), nil
+		}
+	}
+	return 0, nil
+}
+
+type sqlNodeWrapper struct {
```

Comment:
This is probably not a good idea. Fine to wrap a node / row iterator, but actively harmful to pretend to be one. You can't meet the interface requirements in a sane way (panics, non-standard error behavior from Next), so this would die if you actually fed it to an engine to execute instead of iterating over yourself manually. Better to make it explicit that this isn't a Node.